framework/include/ics/InitialConditionBase.h

```diff
@@ -95,47 +95,48 @@ class InitialConditionBase : public MooseObject,
   /**
    * reimplements the getUserObject method from UserObjectInterface
    */
-  template <typename T2>
-  const T2 & getUserObject(const std::string & name);
+  template <typename T>
+  const T & getUserObject(const std::string & name) const;
   /**
    * reimplements the getUserObjectByName method from UserObjectInterface
    */
-  template <typename T2>
-  const T2 & getUserObjectByName(const UserObjectName & name);
+  template <typename T>
+  const T & getUserObjectByName(const UserObjectName & name) const;
 
   /**
    * reimplements the getUserObjectBase method from UserObjectInterface
    */
-  const UserObject & getUserObjectBase(const std::string & name);
+  const UserObject & getUserObjectBase(const std::string & name) const override;
 
 protected:
   /// The system object
   SystemBase & _sys;
 
+  /// If set, UOs retrieved by this IC will not be executed before this IC
+  const bool _ignore_uo_dependency;
+
+private:
   /// Dependent variables
   std::set<std::string> _depend_vars;
   /// Supplied variables
   std::set<std::string> _supplied_vars;
 
   /// Depend UserObjects
-  std::set<std::string> _depend_uo;
-
-  /// If set, UOs retrieved by this IC will not be executed before this IC
-  const bool _ignore_uo_dependency;
+  mutable std::set<std::string> _depend_uo;
```

Comment:
I think this is the proper approach, but would add a detailed comment about why this is mutable.